functions/Invoke-DbaLogShippingRecovery.ps1

```diff
@@ -0,0 +1,307 @@
+
+Recovers all the databases on the instance that are enabled for log shiping using a credential
+
+.EXAMPLE   
+Invoke-DbaLogShippingRecovery -SqlServer 'server1' -database 'db_logship' -Verbose
+
+Recovers the database "db_logship" to a normal status
+
+
+.EXAMPLE   
+db1, db2, db3, db4 | Invoke-DbaLogShippingRecovery -SqlServer 'server1' -Verbose
+
+Recovers the database db1, db2, db3, db4 to a normal status
+
+.EXAMPLE   
+Invoke-DbaLogShippingRecovery -SqlServer 'server1' -Verbose -WhatIf
+
+Shows what would happen if the command were executed.
+
+#>
+    [CmdletBinding(DefaultParameterSetName = "Default", SupportsShouldProcess = $true)]
+    param
+	(
+        [Parameter(Position=1, Mandatory=$true)]
+		[Alias("ServerInstance", "SqlInstance")]
+        [string]$SqlServer,
+        [Parameter(Position=2, Mandatory=$false, ValueFromPipeline=$true)]
+        [string[]]$Database = $null,
+        [Parameter(Position=3, Mandatory=$false)][object]$SqlCredential,
+        [Parameter(Position=4, Mandatory=$false)][switch]$NoRecovery
+	)
+
+    BEGIN
+    {
+        Write-Output "Attempting to connect to Sql Server.."
+		$server = Connect-SqlServer -SqlServer $SqlServer -SqlCredential $SqlCredential
+    }
+
+    PROCESS
+    {
+        Write-Output "Started Log Shipping Recovery."
+
+        # Query for retrieving the log shipping information
+        $query = "
+        SELECT  lss.primary_server ,
+                lss.primary_database ,
+                lsd.secondary_database ,
+                lss.backup_source_directory ,
+                lss.backup_destination_directory ,
+                lss.last_copied_file ,
+                lss.last_copied_date ,
+                lsd.last_restored_file,
+		        sj1.name AS 'copyjob',
+		        sj2.name AS 'restorejob'
+        FROM    msdb.dbo.log_shipping_secondary AS lss
+                INNER JOIN msdb.dbo.log_shipping_secondary_databases AS lsd ON lsd.secondary_id = lss.secondary_id
+		        INNER JOIN msdb.dbo.sysjobs AS sj1 ON sj1.job_id = lss.copy_job_id
+		        INNER JOIN msdb.dbo.sysjobs AS sj2 ON sj2.job_id = lss.restore_job_id "
+        
+        # Check if they commandlet is eecuted from a pipeline
+		if ($PSCmdlet.MyInvocation.ExpectingInput) 
+		{
+			$db = $_
+            
+            $query += "WHERE lsd.secondary_database = '$db'"
+		}
+		else
+		{
+            # If just one database is set in the database parameter
+			if($Database -ne $null)
+            {
+			    $db = $Database[0]
+
+                $query += "WHERE lsd.secondary_database = '$db'"
+            }
+		}
+
+        # Retrieve the log shipping information from the secondary instance
+        try{
+            Write-Verbose "Retrieving log shipping information from the secondary instance"
+            $logshipping_details = Invoke-Sqlcmd2 -ServerInstance $SqlServer -Database 'msdb' -Query $query
+        }
+        catch
+        {
+            Write-Error $_.Exception.Message
+            Break
```

Comment:
It can handle almost all things, but it cannot gracefully and the calling function - that's the limitation of the PowerShell engine's scope system, which means a Stop-Function that is used in a place where it needs to end the calling function cannot do this without throwing an exception (and we wanted to go easy on the red messages).
I tried a good long time getting it to work (something similar to `return return`). Nothing was possible except for using the truly horrid engine statements (such as break) which have seriously bad side effects.
So when we place Stop-Function in a location where we want to actually kill the calling function - as opposed to skip this item and process the next - Stop-Function handles the automatic integrations into anything we want to happen when a function breaks. And if we want to gracefully stop the function the return on the next line does it.